src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -255,29 +268,124 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                 index += 1;
             }
 
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
+            if (Avx2.IsSupported)
             {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<byte>.Count - 1));
+                if ((int)(byte*)index < length)
```

Comment:
Might be good to extract this into an aggressively-inlined helper. It would simplify the overall IL that the JIT sees and make the code more readable.

Reply:
Will get quite messy, shame can't add `[MethodImpl(MethodImplOptions.AggressiveInlining)]` to local functions 😢

Reply:
Done some; inline returns and gotos make the 3 flavours of Vector more difficult to move out to helper calls as have to communicate in the return what to do next; return value, go sequential or return fail; which will probably make the callsite more difficult to optimize after inlining,